### DISCUSSION TITLE
Crazyswarm: frequent crashes and unstable behavior

### DISCUSSION THREAD

**Discussion by GriffinBonner** [USC-ACTLab/crazyswarm#514] [Q&A]
I am currently experiencing an issue with Crazyswarm where when I try to run any of the test scripts (hello_world, niceHover, figure8, etc.) with a single Crazyflie, it works perfectly 20% percent of the time, but otherwise the Crazyflie takes off, quickly becomes unstable and veers off and crashes. This behavior occurs every time I use more than one Crazyflie. I will provide a background on my setup to help with solving this issue.

### Software
Ubuntu 20.04, ROS Noetic, Python 3.7, Crazyswarm, Vicon Tracker 3.4

### Hardware
Crazyflie 2.1, Bitcraze Motion Capture Marker Deck, 6.5 mm markers (5x), 16x Vicon Vantage+ Cameras

### Setup & Configuration
We began by following the crazyswarm installation and configuration tutorials and did not experience any software errors or issues. We plan to use up to ten Crazyflie drones, so each are on the same channel with URI's and names assigned as described in the documentation and enumerated in the .yaml file. The STM/nRF firmware and crazyradio firmware were updated individually. We are using unique marker configurations on each drone with the Vicon motion capture cameras running Vicon Tracker 3.4, so the launch file was edited accordingly. I have attached a photo below of two Crazyflie to demonstrate our marker configurations. The "chooser.py" script to select from a fleet works as described. The scripts run with the simulator argument, the software has not given any errors throughout this setup & configuration process, so I think it has been configured properly. 

### Hardware Testing
The launch file ("hover_swarm.launch") runs without errors and the Crazyflies selected with ("chooser.py") can be seen in rviz. When any of the test scripts are run, more often than not, the Crazyflies takeoff and immediately become unstable and crash. The strange part of this persisting issue is that sometimes the flight is as anticipated and the Crazyflie seems incredibly stable, but this only happens around 20% of the time. I have attached a screen capture of rviz which demonstrates the crashes that I am describing. 

After reviewing a several threads on here, we thought that this may be a latency issue in which the local execution does not have enough time between feedback from the motion capture system to prepare and send packets with the Crazyradio to the Crazyflies. I enabled the verbose latency output option in ("hover_swarm.launch") and have included it below. Based on the feedback given to others on the Issues page, this latency information seems to be sufficient. I am not sure what is causing this issue, is there anything that I have described that is glaringly wrong? Or are there any tips/suggestions on what to try next? This seems to be the same issue that is causing these problems so consistently, and we are unable to proceed with our research without solving this issue. I have also included a rosbag file corresponding to a crash. 

Any feedback or advice is appreciated. 

Best,
Griffin 

### Latency Report ("hover_swarm.launch")
Latencies
Data Delivered: 4.0666 ms
Data Synchronized: 0.1875 ms
Data Processed: 1.3058 ms
Delivered To DataStream: 0.1328 ms
Run All Groups: 0.595091 ms
Group 0 objectTracking: 0 ms
Group 0 broadcasting: 0.256514 ms
Total 6.28779 ms
Avg 23
5.69768,0,0,0,1.87727,0,0.220192,0,0,0,0,7.57495,
Latencies
Data Delivered: 4.0609 ms
Data Synchronized: 0.1732 ms
Data Processed: 1.5954 ms
Delivered To DataStream: 0.1424 ms
Run All Groups: 0.581828 ms
Group 0 objectTracking: 0 ms
Group 0 broadcasting: 0.24213 ms
Total 6.55373 ms
Avg 24
5.7091,0,0,0,1.8233,0,0.221106,0,0,0,0,7.5324,
Latencies
Data Delivered: 4.0149 ms
Data Synchronized: 0.1753 ms
Data Processed: 1.3188 ms
Delivered To DataStream: 0.1385 ms
Run All Groups: 0.625385 ms
Group 0 objectTracking: 0 ms
Group 0 broadcasting: 0.262128 ms
Total 6.27289 ms
Avg 25
5.70664,0,0,0,1.77538,0,0.222747,0,0,0,0,7.48202,

### Latency Report (.txt file)
[latency_output.txt](https://github.com/USC-ACTLab/crazyswarm/files/7437336/latency_output.txt)

### Crazyflie Marker Configuration (Unique)
![IMG_20211028_162230](https://user-images.githubusercontent.com/34168079/139330162-1f3b7af7-cc49-4cda-bf7c-2583a0e4cb43.jpg)

### RVIZ Crash
https://user-images.githubusercontent.com/34168079/139334647-02179e99-42d3-4c0d-aae5-b0f7bcc2ef9b.mp4

### ROS bag file (crash)
[cf2_10_hover.zip](https://github.com/USC-ACTLab/crazyswarm/files/7437340/cf2_10_hover.zip)

**Comment by jpreiss** [comment]
Latency numbers look fine.

This kind of immediate instability is often related to the object tracking. In `latency_output.txt` it looks like you have 4 CFs, but in the video there is only one. How many CFs are in the scene?

If there are 4 CFs in the scene, this means Vicon Tracker is tracking all 4, even if they are not all flying. I have observed Vicon Tracker swap the identities of two CFs with unique marker configurations even though the configurations are reasonably different. Unfortunately I never resolved the issue, but Vicon Tracker has many settings to adjust.

Does the problem still occur if you fly with only one CF in the scene? (Make sure to disable all other objects in Vicon Tracker too.)

**Reply by whoenig** [reply]
Just to add here: Which motion capture mode did you use (in your case we'd recommend libobjecttracker)? Which firmware did you end up flashing? Do you have a LED ring to verify that (broadcast) communication is working?

**Reply by GriffinBonner** [reply]
I am using "motionCapture" which I thought was the right option when using Vicon and unique marker configurations. I am using the most recent firmware per the documentation instructions. I do not have a LED ring, what do you recommend testing in terms of "broadcast communication" ? @jpreiss I should have clarified this in the post, but the rviz recording and the latency output are from two different trials, the rviz is from a trial with a single crazyflie and the latency output is from a trial with a single crazyflie. Why should I be using libobjecttracker if I have unique marker configurations, I thought that was for duplicated marker configurations.

**Reply by whoenig** [reply]
You can enable logging and pose logging in the launch file. Then look at /cfID/pose using e.g., rostopic echo, and see if the values change correctly as you move the drone around manually.

Yes, if you want vicon to track your drones, the correct setting is `motioncapture`.

**Reply by GriffinBonner** [reply]
I will check this as well, but I think this is working as expected because I was monitoring the /tf rostopic and it seemed to be giving accurate transformations of the crazyflie positions from the world coordinate system. Are there any common issues that people have reported immediate instability have resolved? The Vicon Tracker is running at 100 Hz and the frequency of /tf is (number of crazyflies) * 100 Hz. It seems like it is some type of tracking issue, but I do not know what else to troubleshoot.

**Reply by whoenig** [reply]
/tf is different, because it shows you the mocap data. The `/cfID/pose` topic shows you what the drone actually estimates using the on-board EKF.

There are plenty of reports like that, but mostly related to a wrong coordinate systems with Optitrack.

**Reply by GriffinBonner** [reply]
Is there a place to find all of the launch file parameters? I cannot figure out how to enable "pose logging" in the launch file, I was able to enable logging, but I am still unable to see the /cfID/pose topics. I have ensured that only crazyflies that are enabled in Vicon tracker are selected and present in the motion capture area, but even with a single crazyflie I still experience the same initial instability.

**Reply by whoenig** [reply]
https://github.com/USC-ACTLab/crazyswarm/blob/27ef265cb0dfa910abfa67705724b8acffb20eaf/ros_ws/src/crazyswarm/launch/hover_swarm.launch#L44

There is no automatic way of listing all parameters, since some of them have default values.

**Reply by GriffinBonner** [reply]
Ok, thank you. One more question, how would you recommend that I troubleshoot if the crazyflie is actually receiving the motion capture position information? This seems like the most likely problem, I know that it is easy to troubleshoot with the LED light, but I do not currently have one available. I am able to see the proper positions of the crazyflie in rviz (as seen in posted video), but I didn't know until reading several other discussions that it is possible that the crazyflies is never receiving this information.

**Reply by whoenig** [reply]
rviz shows you the belief of the host computer. The `cfID/pose` topic shows you the belief of the drone itself. If they don't match, then there is a communication issue. (Verify that they match while manually walking the drone around.)

**Reply by GriffinBonner** [reply]
I confirmed that the cfID/pose topic matches the /tf topic, and the ROS coordinate system matches the motion capture coordinate system.

**Reply by whoenig** [reply]
And you see the same crash as initially posted? Which tracking mode are you using (libobjecttracker or motioncapture)? If you use the VICON Tracker software, make sure to have the motion filters enabled to avoid any spurious axis flips.

**Reply by GriffinBonner** [reply]
It is the same crash, but it happens roughly 70% of the time with a single crazyflie, and 100% of the time when adding two or more crazyflies. I am using motionCapture, I will enable that setting in VICON tracker and see what happens.
